tests/test_FX_revaluation.py

```diff
@@ -0,0 +1,108 @@
+"""
+
+# flake8: enable
+
+STRIPPED_CSV = "\n".join([line.strip() for line in LEDGER_CSV.split("\n")])
+LEDGER_ENTRIES = pd.read_csv(
+    StringIO(STRIPPED_CSV), skipinitialspace=True, comment="#", skip_blank_lines=True
+)
+TEST_ACCOUNTS = pd.read_csv(StringIO(ACCOUNT_CSV), skipinitialspace=True)
+
+
+@pytest.fixture(scope="session")
+def set_up_ledger_and_account():
+    cashctrl = CashCtrlLedger()
+    cashctrl.transitory_account = 1000
+
+    # Fetch original state
+    initial_account_chart = cashctrl.account_chart().reset_index()
+    initial_ledger = cashctrl.ledger()
+
+    # Create test accounts and Ledger
+    cashctrl.mirror_account_chart(TEST_ACCOUNTS, delete=False)
+    cashctrl.mirror_ledger(LEDGER_ENTRIES, delete=False)
+
+    yield
+
+    # Restore initial state
+    cashctrl.mirror_ledger(initial_ledger, delete=True)
+    cashctrl.mirror_account_chart(initial_account_chart, delete=True)
+
+
+def test_FX_revaluation(set_up_ledger_and_account):
+    cashctrl = CashCtrlLedger()
+    accounts = pd.DataFrame({
+        "foreign_currency_account": [10001, 10002, 10004],
+        "fx_gain_loss_account": [10003, 10003, 10003],
+        "exchange_rate": [0.75, 0.5, None],
+    })
+    eur_account_id = cashctrl._client.account_to_id(10001)
+    usd_account_id = cashctrl._client.account_to_id(10002)
+    usd_na_account_id = cashctrl._client.account_to_id(10004)
+
+    cashctrl.FX_revaluation(accounts=accounts)
+    ex_diff = cashctrl._client.get("fiscalperiod/exchangediff.json")["data"]
+    eur_account = next((item for item in ex_diff if item['accountId'] == eur_account_id), None)
+    usd_account = next((item for item in ex_diff if item['accountId'] == usd_account_id), None)
+    usd_na_account = next((item for item in ex_diff if item['accountId'] == usd_na_account_id), None)
+
+    from_currency = cashctrl._client.account_to_currency(10004)
+    params = {"from": from_currency, "to": cashctrl.base_currency, "date": None}
+    response = cashctrl._client.request("GET", "currency/exchangerate", params=params)
+    usd_na_account_fx_rate = response.json()
+    usd_na_account_dcBalance = round(100 * usd_na_account_fx_rate, 2)
+
+    assert eur_account is not None, "EUR account not found in exchange differences"
+    assert usd_account is not None, "USD account not found in exchange differences"
+    assert usd_na_account is not None, "USD account with NA fx rate not found in exchange differences"
+    assert eur_account['dcBalance'] == 75.0, (
+        f"EUR account dcBalance is {eur_account['dcBalance']}, expected 50.0")
+    assert usd_account['dcBalance'] == 50.0, (
+        f"USD account dcBalance is {usd_account['dcBalance']}, expected 50.0")
+    assert usd_na_account['dcBalance'] == usd_na_account_dcBalance, (
+        f"USD account dcBalance is {usd_na_account['dcBalance']}, expected {usd_na_account_dcBalance}")
```

Comment:
Readability of these tests could be vastly improved once above suggestion to implement an `FX_valuation` function is implemented.,